packages/meditrak-server/src/database/SurveyResponseChangeHandler/OutdatedStatusUpdater.js

```diff
@@ -0,0 +1,260 @@
+ */
+
+const DATE_FORMAT = 'YYYY-MM-DD';
+const DATETIME_FORMAT = `${DATE_FORMAT} HH:mm:ss`;
+
+/**
+ * Cache of period ranges keyed by requested period granularity first, then date
+ *
+ * ```js
+ * {
+ *   yearly: {
+ *     "2021-05-05": { startDate: '2021-01-01', endDate: '2021-12-31' }
+ *   }
+ *   monthly: {
+ *     "2021-07-06": { startDate: '2021-07-01', endDate: '2021-07-31' }
+ *   }
+ * }
+ * ```
+ */
+const periodRangeCache = Object.fromEntries(
+  Object.values(PERIOD_GRANULARITIES).map(granularity => [granularity, {}]),
+);
+
+/**
+ * @param {Date|string} datetime
+ * @returns {string}
+ */
+const extractDateString = datetime => {
+  const datetimeString = typeof datetime === 'string' ? datetime : datetime.toISOString();
+  return datetimeString.substring(0, DATE_FORMAT.length);
+};
+
+/**
+ */
+const getPeriodRange = (periodGranularity, datetime) => {
+  // None of the existing granularities depends on time, so we can just use the date part
+  // to cache/retrieve date ranges
+  const date = extractDateString(datetime);
+  const existingRange = periodRangeCache?.[periodGranularity]?.[date];
+  if (existingRange) {
+    return existingRange;
+  }
+
+  const momentUnit = PERIOD_GRANULARITY_TO_MOMENT_UNIT[periodGranularity];
+  const range = {
+    startDate: moment(datetime).startOf(momentUnit).format(DATETIME_FORMAT),
+    endDate: moment(datetime).endOf(momentUnit).format(DATETIME_FORMAT),
+  };
+  periodRangeCache[periodGranularity][date] = range;
+
+  return range;
+};
+
+const getIdOfMostRecentResponse = surveyResponses => {
+  assert.notStrictEqual(surveyResponses.length, 0);
+
+  const comparableResponses = surveyResponses.map(response => ({
+    ...response,
+    // If other ordering criteria are the same, not outdated responses take precedence over outdated
+    // Assign numeric value for comparison
+    notOutdated: response.outdated ? 0 : 1,
+  }));
+
+  return orderBy(
+    comparableResponses,
+    ['end_time', 'notOutdated', 'id'],
+    ['desc', 'desc', 'desc'],
+  )[0].id;
+};
+
+const getDimensionCombo = (surveyResponse, survey) => {
+  const { period_granularity: periodGranularity } = survey;
+  const { data_time: dataTime } = surveyResponse;
+  const { startDate, endDate } = getPeriodRange(periodGranularity, dataTime);
+
+  return {
+    surveyId: survey.id,
+    entityId: surveyResponse.entity_id,
+    startDate,
+    endDate,
+  };
+};
+
+const groupResponsesForSurveyByDimensionCombo = (surveyResponses, survey) =>
+  Object.values(groupBy(surveyResponses, sr => JSON.stringify(getDimensionCombo(sr, survey))));
+
+export class OutdatedStatusUpdater {
+  constructor(models) {
+    this.models = models;
+  }
+
+  /**
+   * Not all changed survey responses need to trigger an outdated status update
+   * This method can be used to filter out changed records that are not worth
+   * considering, without having to look into the DB for more information
+   */
+  static getCandidatesFromChangeDetails = changeDetails => {
```

Comment:
Absolutely love how clever this function is to filter changes and only process the necessary ones! I think we should do something similar for the `AnalyticsRefresher` too!

Reply:
Yeap good call!